### PR TITLE
Add support for `pandas-2.2` in `cudf`

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
This PR:

- [x] Enables `pandas-2.2` in `cudf` by upgrading the upper bound pinnings.
- [x] Cleans up a lot of dead-code.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/column/datetime.py
Comment (mroeschke) [inline python/cudf/cudf/core/column/datetime.py:328]: Just in noting in the future we should pass pandas a numpy array instead of a pyarrow array and `pandas.Series` treats pyarrow arrays as objects

**Review by mroeschke — COMMENTED** [review]
If/when we decide to test against the minimum supported pandas version (pandas 2.0) would we have to add any of these checks back in?

**Comment by galipremsagar** [comment]
> If/when we decide to test against the minimum supported pandas version (pandas 2.0) would we have to add any of these checks back in?

The current `PANDAS_GE_200` code deletions will also hold if we want to test with pandas-2.0 (lowest supported version), but I think it's highly unlikely we can keep up and maintain such a test-suite in CI.

**Review by mroeschke — APPROVED** [review]

File: conda/environments/all_cuda-118_arch-x86_64.yaml
Comment (shwina) [inline conda/environments/all_cuda-118_arch-x86_64.yaml:68]: Is this the right pinning, or should we go with `<2.3.0dev0`? Do we expect patch releases to break us?

Comment (galipremsagar) [reply conda/environments/all_cuda-118_arch-x86_64.yaml:68]: I feel various fixes are going into `2.2.1` that may have an impact on us, but would defer to @mroeschke on that.

**Review by shwina — APPROVED** [review]
Approving with a question about pinnings that we should resolve before merging.

File: python/cudf/cudf/core/_compat.py
Comment (bdice) [inline python/cudf/cudf/core/_compat.py:13]: Can we sort these, so we have all the `EQ` then `GE` then `LT`? It looks like you deleted a duplicate line but it ended up breaking up the groupings.

**Review by bdice — APPROVED** [review]

File: python/cudf/cudf/core/_compat.py
Comment (galipremsagar) [reply python/cudf/cudf/core/_compat.py:13]: Yup, I'll do that in a follow-up pr. Potentially there is still some cleanup to be done.

File: conda/environments/all_cuda-118_arch-x86_64.yaml
Comment (mroeschke) [reply conda/environments/all_cuda-118_arch-x86_64.yaml:68]: Yeah I fixed various issues @galipremsagar found in pandas 2.2.1 (coming out next week) which there are xfails for so the update will break the test suite. We should probably relax the pin in a follow up PR when that's out

**Review by raydouglass — APPROVED** [review]

File: ci/cudf_pandas_scripts/pandas-tests/run.sh
Comment (galipremsagar) [inline ci/cudf_pandas_scripts/pandas-tests/run.sh:35]:
```suggestion
  -n 10 -v \
```

File: python/cudf/cudf/pandas/scripts/run-pandas-tests.sh
Comment (galipremsagar) [inline python/cudf/cudf/pandas/scripts/run-pandas-tests.sh:25]:
```suggestion
PYTEST_IGNORES="--ignore=tests/io/test_user_agent.py --ignore=tests/interchange/test_impl.py --ignore=tests/window/test_dtypes.py --ignore=tests/strings/test_api.py"
```

**Comment by galipremsagar** [comment]
@shwina @raydouglass @mroeschke In the interest of time, I upgraded the pinnings to include the latest patch release of pandas-2.2.1 too. If you have any concerns on these changes, let me know I'll address it in a follow-up pr.

**Pull request merged**
